### DISCUSSION TITLE
Custom component development—pyOCD plugins

### DISCUSSION THREAD

**Discussion by diggit** [pyocd/pyOCD#1367] [General]
Hi,
Let's say I want to add support for another RTOS, but I don't want to edit pyOCD sources directly. Is pyOCD able to load external components?
e.g. pemicro probe is standalone python package, so I guess it should be possible...

**Comment by flit** [comment]
Yup! To some extent, at least. Pyocd currently supports RTOS and debug probe plugins. So it sounds like you should be ok.

(I'm converting this issue to a discussion and will continue with more details there.)

**Comment by diggit** [comment]
Perfect!
Unless there are docs about it, could you nudge me to the correct direction? Eg. where pyOCD looks for these plugins or how to tell it which to load.

**Comment by flit** [comment]
Pyocd uses setuptools [entry points](https://setuptools.pypa.io/en/latest/userguide/entry_point.html) for its plugins. All of the built-in RTOSes and debug probes are also defined and accessed as plugins, so you can see how they are defined in the [`[options.entry_points]` section](https://github.com/pyocd/pyOCD/blob/88222188de5ee553ada083efbdd54eb8f15974b0/setup.cfg#L81) of `setup.cfg`. So, to define an external plugin you need to create a Python package or module that exports the relevant entry point(s).

The two entry points that pyocd currently uses are `pyocd.probe` and `pyocd.rtos`. The entry point must reference a [`pyocd.core.plugin.Plugin`](https://github.com/pyocd/pyOCD/blob/88222188de5ee553ada083efbdd54eb8f15974b0/pyocd/core/plugin.py#L34) subclass. This subclass will have some properties with the plugin name, description, any options. The `.load()` method must return the type-specific implementation class. For RTOS plugins that's a subclass of [`pyocd.rtos.provider.ThreadProvider`](https://github.com/pyocd/pyOCD/blob/88222188de5ee553ada083efbdd54eb8f15974b0/pyocd/rtos/provider.py#L47).

The [`FreeRTOSPlugin`](https://github.com/pyocd/pyOCD/blob/88222188de5ee553ada083efbdd54eb8f15974b0/pyocd/rtos/freertos.py#L536) class is a simple reference example.
